### DISCUSSION TITLE
DynamoDBのデータ量を減らしたい

### DISCUSSION THREAD

**Discussion by koichimurakami** [mbc-net/mbc-cqrs-serverless#48] [Ideas]
現在DynamoDBにコマンド、最新データ、履歴データが保存されています。
なるべくコストを抑えるためにデータ量を減らすための設定が出来るようにしてほしいです。

## 設定データ種別
* コマンド履歴データの保有期間: デフォルト 90
* 最新データの保有期間: デフォルトnull
* 履歴データの保有期間: デフォルト 90

いずれの項目も null は設定可能としてください。null が設定された際は無制限に保存するものとします。

## DynamoDBの保有期間
CDKで設定データを定義して、DynamoDBのTTLに反映するようにして下さい。

※ 設定変更をした場合 DynamoDB TTLに変更しません。保管期間を変更したい場合は別途対応が必要とします。

### コマンド
最新のコマンド以外は `コマンド履歴データの保有期間` に基づき自動的に削除されるようにしてください。
最新のコマンドについては `最新データの保有期間` を過ぎたら削除するようにしてください。

最新コマンドが追加された際の処理として、 `最新データの保有期間` を最新コマンドに設定し、古いコマンドについて `コマンド履歴データの保有期間` を自動セットする必要があります。

### 最新データ
`最新データの保有期間` を過ぎたら削除するようにしてください。

### 履歴データ
`履歴データの保有期間`  を過ぎたら削除するようにしてください。

**Comment by baquan1708** [comment]
@koichimurakami cc: @Tuan-Anh-Nguyen-pionero 

In the current deployment flow, we create DDB and RDS tables by running the `npm run migrate` command before deploying the CDK stack. The `npm run migrate` script includes the `npm run migrate:ddb` command, which performs the following tasks for DDB tables:

- Creates DDB tables.
- Enables the stream feature for command-type tables.
- Enables TTL and point-in-time recovery features for all table types.

**Note**: To define a data retention period in DynamoDB, the TTL attribute must be set during item creation.

The solution for this requirement involves the following:

- Add 3 environment variables to specify retention periods for different data types (command, history, data) and apply TTL for each type of data during the publishing and processing of commands.
- Add a `set_ttl_command` step before the history_copy step to set TTL for older commands.
- Modify the `history_copy` step to set TTL for the history data.

Could you please review the solution  before I proceed with the implementation?

![image](https://github.com/user-attachments/assets/c5c903f1-312f-48c8-af7c-eb8c7362aa5c)

**Reply by koichimurakami** [reply]
@baquan1708  cc: @Tuan-Anh-Nguyen-pionero 

Is it possible to specify the data retention period for each table?

For example, the retention periods differ for master tables, estimate tables, and sales tables.  Since the data retention period differs depending on the characteristics of each table, we cannot uniformly determine the retention period for command tables, data tables, and history tables.

Furthermore, there is a possibility that requirements to change the retention period may arise during operation.

While `npm run migrate` command would be acceptable, could you please consider allowing the retention period to be set for each table? 

I understand that `set_ttl_command` command needs to be executed, but if the retention period setting is available within the framework, can it be executed automatically? (If possible, we would like to avoid adding it to the user's program.)

**Reply by baquan1708** [reply]
@koichimurakami cc: @Tuan-Anh-Nguyen-pionero 

Yes, it is possible to specify the retention period (TTL) for each table. In this case, we can store separate TTL settings for each table into DDB table (e.g. store in master table), allowing us to modify the TTL dynamically during operation.

There is, however, a constraint for TTL to ensure asynchronous processing works correctly: 
- `最新データの保有期間` >= `コマンド履歴データの保有期間`.

Thus, there are two TTL configurations:
- **TTL from DDB configuration:** If we change the TTL, additional actions will be required, such as applying the new TTL to old commands and history items.
- **Default TTL:** stored indefinitely.

If the "TTL from DDB configuration" exists, it will be used; otherwise, the "default TTL" will be applied.

--- 

Changes in the Step Function, like adding a `set_ttl_command` step or modifying the `history_copy` step, will be executed automatically within the framework. Therefore, the users do not need to write any additional code.

**Reply by koichimurakami** [reply]
@baquan1708 cc: @Tuan-Anh-Nguyen-pionero 
Thank you.
So, the DynamoDB data looks good from the above.
By the way, if the command data volume is large, I think the framework is saving some of the attributes to S3, but do you have any good ideas for S3 as well?

**Reply by baquan1708** [reply]
@koichimurakami cc: @Tuan-Anh-Nguyen-pionero 

Thank you for bringing this to my attention.

If the command data size is large, the framework will store the attribute to s3 with key: `ddb/${tableName}/${pk}/${sk}/${ulid()}.json`. Therefore, if we change the TTL, additional actions will be required, such as 
- applying the new TTL to old commands and history items, 
- **and set the bucket lifecycle configuration for `ddb/ddb/${tableName}`**.

Note: An S3 Lifecycle configuration can have up to 1,000 rules. This limit is not adjustable. [Ref](https://docs.aws.amazon.com/AmazonS3/latest/userguide/intro-lifecycle-rules.html#intro-lifecycle-rule-id).

**Reply by koichimurakami** [reply]
@baquan1708 cc @Tuan-Anh-Nguyen-pionero 
Thank you for reply. Limitation of lifecycle configuration is not so important. Project may be micro service.

Could you please proceed with the implementation

**Reply by baquan1708** [reply]
@koichimurakami cc: @Tuan-Anh-Nguyen-pionero 
Yes, I will implement this right away.

**Reply by baquan1708** [reply]
@koichimurakami cc: @Tuan-Anh-Nguyen-pionero 

I've just created PR: https://github.com/mbc-net/mbc-cqrs-serverless/pull/97. When you have a moment, could you please review it?

Here is the step function diagram after the change:

![image](https://github.com/user-attachments/assets/b5858b80-da52-4ca9-9ac3-3ef644d76983)

**Comment by koichimurakami** [comment]
Release Info in Japanese https://www.mbc-net.com/update-mbc-cqrs-%e3%82%b5%e3%83%bc%e3%83%90%e3%83%bc%e3%83%ac%e3%82%b9-%e3%83%95%e3%83%ac%e3%83%bc%e3%83%a0%e3%83%af%e3%83%bc%e3%82%af-ver-0-1-44-beta-dynamodb%e3%81%ae%e3%83%87%e3%83%bc%e3%82%bf/
